fendermint/vm/interpreter/src/fvm/observe.rs

```diff
@@ -18,63 +20,40 @@ register_metrics! {
         = register_histogram!("exec_fvm_call_execution_time_secs", "Execution time of FVM call in seconds");
 }
 
-impl_traceables!(
-    TraceLevel::Info,
-    "Execution",
-    MsgExecCheck<'a>,
-    MsgExecEstimate<'a>,
-    MsgExecApply<'a>,
-    MsgExecCall<'a>
-);
-
-macro_rules! message_exec_struct {
-    ($($name:ident),*) => {
-        $(
-            #[derive(std::fmt::Debug)]
-            #[allow(dead_code)]
-            pub struct $name<'a> {
-                pub height: i64,
-                pub from: &'a str,
-                pub to: &'a str,
-                pub value: &'a str,
-                pub method_num: u64,
-                pub gas_limit: u64,
-                pub gas_price: &'a str,
-                pub params: &'a [u8],
-                pub nonce: u64,
-                pub duration: f64,
-                pub exit_code: u32,
-            }
-        )*
-    };
-}
-
-impl Recordable for MsgExecCheck<'_> {
-    fn record_metrics(&self) {
-        EXEC_FVM_CHECK_EXECUTION_TIME_SECS.observe(self.duration);
-    }
-}
+impl_traceables!(TraceLevel::Info, "Execution", MsgExec);
 
-impl Recordable for MsgExecEstimate<'_> {
-    fn record_metrics(&self) {
-        EXEC_FVM_ESTIMATE_EXECUTION_TIME_SECS.observe(self.duration);
-    }
+#[derive(Debug, strum::EnumString)]
+#[strum(serialize_all = "snake_case")]
+pub enum MsgExecPurpose {
+    Check,
+    Apply,
+    Estimate,
+    Call,
 }
 
-impl Recordable for MsgExecApply<'_> {
-    fn record_metrics(&self) {
-        EXEC_FVM_APPLY_EXECUTION_TIME_SECS.observe(self.duration);
-    }
+#[derive(Debug)]
+#[allow(dead_code)]
+pub struct MsgExec {
+    pub purpose: MsgExecPurpose,
+    pub message: Message,
+    pub height: i64,
+    pub duration: f64,
+    pub exit_code: u32,
 }
 
-impl Recordable for MsgExecCall<'_> {
+impl Recordable for MsgExec {
     fn record_metrics(&self) {
-        EXEC_FVM_CALL_EXECUTION_TIME_SECS.observe(self.duration);
+        match self.purpose {
+            MsgExecPurpose::Check => EXEC_FVM_CHECK_EXECUTION_TIME_SECS.observe(self.duration),
+            MsgExecPurpose::Estimate => {
+                EXEC_FVM_ESTIMATE_EXECUTION_TIME_SECS.observe(self.duration)
+            }
+            MsgExecPurpose::Apply => EXEC_FVM_APPLY_EXECUTION_TIME_SECS.observe(self.duration),
+            MsgExecPurpose::Call => EXEC_FVM_CALL_EXECUTION_TIME_SECS.observe(self.duration),
+        }
     }
 }
```

Comment:
Nice! It's all coming together. This would've been excrutiatingly difficult to do with the previous code.

Reply:
Yes, it is also much cleaner this way.